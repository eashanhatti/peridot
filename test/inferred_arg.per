/* Inferred arguments. The typechecker figures out the value of an inferred argument when the function is called, based on the types of the other arguments. Because `x : A`, the typechecker will know the value of `A` once the type of `x` is known */
define id
  : Function(inferred A: Type, A) -> A
  = function(x) => x;

/* `x = true`, so the typechecker knows `A = Bool` */
define x: Bool = id(true);

/* Inferred arguments can also be explicitly pass with the `explicit` keyword */
define y: Bool = id(explicit Bool, true);
