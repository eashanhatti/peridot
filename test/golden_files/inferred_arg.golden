((fromList [(UVGlobal 1000, Just (Rigid TwoType))],
  fromList [],
  fromList []),
 Let
   (fromList
      [ObjTerm
         (Id {unId = 0})
         (ObjFunType
            Unification
            (Rigid (TypeType Obj))
            (ObjFunType
               Explicit
               (LocalVar (Index {unIndex = 0}))
               (LocalVar (Index {unIndex = 1}))))
         (ObjFunIntro
            (ObjFunIntro (LocalVar (Index {unIndex = 0})))),
       ObjTerm
         (Id {unId = 1})
         (Rigid TwoType)
         (ObjFunElim
            (ObjFunElim
               (GlobalVar (Id {unId = 0})) (UniVar (UVGlobal 1000)))
            (Rigid TwoIntro0)),
       ObjTerm
         (Id {unId = 2})
         (Rigid TwoType)
         (ObjFunElim
            (ObjFunElim
               (GlobalVar (Id {unId = 0})) (Rigid TwoType))
            (Rigid TwoIntro0))])
   (Rigid (TypeType Obj)))
